File: src/pensiontracker/formatters.py
"""
formatters.py
-------------
Parsing y formateo de números en el formato chileno (miles con puntos,
decimales con coma). Compartido entre blueprints (routes/pagos.py,
routes/utm.py).
"""

import math
import re

_FACTOR_VALIDO = re.compile(r"^-?[0-9]*\.?[0-9]+$")
_ENTERO_VALIDO = re.compile(r"^-?[0-9]+$")


def limpiar_entero(valor: str) -> int:
    """
    Convierte texto formateado en Chile (miles con puntos) a entero.
    Ej: '69.889' → 69889 | '213.588' → 213588 | '1000' → 1000
    Rechaza comas (separador decimal no permitido para enteros).

    int() de Python acepta formas que no son enteros chilenos válidos:
    signo '+' explícito ('+69889'), guión bajo como separador de dígitos
    ('1_000') y dígitos Unicode no ASCII (p. ej. arábigo-índicos). El
    TypeScript ya rechaza las tres con el regex ^-?\\d+$ (donde \\d sin
    flag /u solo matchea 0-9); acá se exige la misma forma estricta
    ([0-9] explícito, no \\d, que en un patrón Python sí matchea dígitos
    Unicode) antes de convertir, para que ambos lados rechacen igual.
    """
    if not valor:
        raise ValueError("Valor vacío")
    limpio = valor.strip()
    if "," in limpio:
        raise ValueError("No se permiten decimales en este campo")
    limpio = limpio.replace(".", "")
    if not _ENTERO_VALIDO.match(limpio):
        raise ValueError(f"Entero inválido: {valor!r}")
    return int(limpio)


def limpiar_factor(valor: str) -> float:
    """
    Convierte texto de factor UTM a float.

    Acepta punto o coma como separador decimal: en los teclados decimales
    de celular el punto suele ser lo único disponible, así que rechazarlo
    convertiría '3.5' en 35 y decuplicaría la cuota pactada.

    Regla: el último separador presente es el decimal; los anteriores son
    separadores de miles y se descartan.

    Ej: '3,0561' → 3.0561 | '3.0561' → 3.0561 | '1.234,56' → 1234.56
    """
    if not valor:
        raise ValueError("Valor vacío")

    limpio = valor.strip()
    if not limpio:
        raise ValueError("Valor vacío")

    # Un separador final sin dígitos detrás (ej. '3,5.') no es el decimal:
    # tratarlo como tal dejaría la parte decimal vacía y descartaría el
    # separador anterior, que sí lo era, decuplicando el resultado. Se
    # ignora y el separador previo (si existe) sigue siendo el decimal.
    while limpio and limpio[-1] in ".,":
        limpio = limpio[:-1]
    if not limpio:
        raise ValueError(f"Factor UTM inválido: {valor!r}")

    corte = max(limpio.rfind("."), limpio.rfind(","))
    if corte == -1:
        entero, decimales = limpio, ""
    else:
        entero, decimales = limpio[:corte], limpio[corte + 1:]

    entero = entero.replace(".", "").replace(",", "")
    normalizado = f"{entero}.{decimales}" if decimales else entero

    # float() acepta notación científica ('1e10'), signo '+' y 'nan'/'inf',
    # pero un factor UTM no admite ninguna de esas formas en la realidad.
    # Se exige la misma forma estricta que el TypeScript (^-?\d*\.?\d+$,
    # donde \d sin flag /u solo matchea 0-9); acá se usa [0-9] explícito
    # en vez de \d, que en un patrón Python sí matchea dígitos Unicode
    # (p. ej. arábigo-índicos), para que ambos lados rechacen igual.
    if not _FACTOR_VALIDO.match(normalizado):
        raise ValueError(f"Factor UTM inválido: {valor!r}")

    try:
        resultado = float(normalizado)
    except ValueError:
        raise ValueError(f"Factor UTM inválido: {valor!r}")

    if not math.isfinite(resultado):
        raise ValueError(f"Factor UTM inválido: {valor!r}")

    return resultado


def fmt_factor(n):
    """Formatea un factor UTM para el frontend: '3.0561' → '3,0561' (sin ceros de más)."""
    if n is None:
        return ""
    s = f"{float(n):.4f}".replace(".", ",")
    return s.rstrip("0").rstrip(",") if "," in s else s

File: mobile/src/core/formatters.ts
/**
 * Parsing y formateo de números en formato chileno: miles con punto,
 * decimales con coma.
 *
 * Port de src/pensiontracker/formatters.py. Debe mantenerse en paridad
 * con esa implementación; las fixtures doradas verifican ambas.
 *
 * ## Diferencias conocidas y deliberadas
 *
 * - `fmtFactor(-0)` da `"0"` aquí y `"-0"` en Python: inalcanzable en la
 *   práctica porque el factor UTM siempre se valida positivo (> 0) antes
 *   de llegar a formatearse, así que nunca se formatea un -0.
 * - `redondear(NaN)` / `redondear(Infinity)` lanzan excepción aquí,
 *   mientras que `round()` de Python los devuelve tal cual: el
 *   guardarraíl es intencional y también inalcanzable, porque
 *   `limpiarFactor` rechaza los valores no finitos antes de que lleguen
 *   a redondear.
 */

import { redondear } from './redondeo';

/**
 * Convierte texto de factor UTM a número.
 *
 * Acepta punto o coma como separador decimal: en los teclados decimales
 * de celular el punto suele ser lo único disponible.
 *
 * Regla: el último separador presente es el decimal; los anteriores son
 * separadores de miles y se descartan. Un separador final suelto se
 * ignora, porque no delimita ninguna parte decimal.
 */
export function limpiarFactor(valor: string): number {
  if (!valor) {
    throw new Error('Valor vacío');
  }

  let limpio = valor.trim();
  while (limpio.length > 0 && (limpio.endsWith('.') || limpio.endsWith(','))) {
    limpio = limpio.slice(0, -1);
  }
  if (limpio.length === 0) {
    throw new Error('Valor vacío');
  }

  const corte = Math.max(limpio.lastIndexOf('.'), limpio.lastIndexOf(','));
  let entero: string;
  let decimales: string;
  if (corte === -1) {
    entero = limpio;
    decimales = '';
  } else {
    entero = limpio.slice(0, corte);
    decimales = limpio.slice(corte + 1);
  }

  entero = entero.replaceAll('.', '').replaceAll(',', '');
  const normalizado = decimales ? `${entero}.${decimales}` : entero;

  // Number() acepta cadena vacía como 0 y tolera espacios; se exige que
  // la cadena sea exactamente un número decimal.
  if (!/^-?\d*\.?\d+$/.test(normalizado)) {
    throw new Error(`Factor UTM inválido: ${JSON.stringify(valor)}`);
  }

  const resultado = Number(normalizado);
  if (!Number.isFinite(resultado)) {
    throw new Error(`Factor UTM inválido: ${JSON.stringify(valor)}`);
  }
  return resultado;
}

/**
 * Convierte texto formateado en Chile (miles con puntos) a entero.
 * Rechaza comas: en este campo no se permiten decimales.
 */
export function limpiarEntero(valor: string): number {
  if (!valor) {
    throw new Error('Valor vacío');
  }
  const limpio = valor.trim();
  if (limpio.includes(',')) {
    throw new Error('No se permiten decimales en este campo');
  }
  const sinPuntos = limpio.replaceAll('.', '');
  if (!/^-?\d+$/.test(sinPuntos)) {
    throw new Error(`Entero inválido: ${JSON.stringify(valor)}`);
  }
  return Number(sinPuntos);
}

/** Formatea un factor UTM para mostrar: 3.0561 → '3,0561', sin ceros de más. */
export function fmtFactor(n: number | null | undefined): string {
  if (n === null || n === undefined) {
    return '';
  }
  const s = n.toFixed(4).replace('.', ',');
  return s.includes(',') ? s.replace(/0+$/, '').replace(/,$/, '') : s;
}

/** Formatea un monto como moneda chilena: 68923.5 → '$68.924'. */
export function formatearPesos(monto: number): string {
  // redondear() mira la expansión decimal exacta del double en vez de
  // restar el piso y comparar contra 0.5: ese atajo fabrica empates que
  // no existen (ver el comentario de redondeo.ts) y hace divergir el
  // resultado de Python en valores como 1.4999999999999998.
  const entero = redondear(monto, 0);
  // El signo sale del monto original y no del entero redondeado: Python
  // conserva el "-" aunque la magnitud redondee a cero ("$-0"), y la
  // paridad con el escritorio manda sobre la estética.
  const signo = monto < 0 || Object.is(monto, -0) ? '-' : '';
  const digitos = Math.abs(entero).toString();
  const conPuntos = digitos.replace(/\B(?=(\d{3})+(?!\d))/g, '.');
  return `$${signo}${conPuntos}`;
}
